File: HealthLink_FE/src/components/patient-dashboard/pharmacy-chat-policy.test.mjs
import assert from 'node:assert/strict';
import test from 'node:test';
import { getPatientPharmacyChatMode } from './pharmacy-chat-policy.js';

test('getPatientPharmacyChatMode returns editable for IN_REVIEW consultation without order', () => {
  assert.equal(getPatientPharmacyChatMode({
    request: { requestType: 'CONSULTATION', status: 'IN_REVIEW', chatRoomId: 'room-41' },
    order: null,
  }), 'editable');
});

test('getPatientPharmacyChatMode returns editable for REVISION_REQUESTED order', () => {
  assert.equal(getPatientPharmacyChatMode({
    request: { requestType: 'CONSULTATION', chatRoomId: 'room-41' },
    order: { status: 'REVISION_REQUESTED' },
  }), 'editable');
});

test('getPatientPharmacyChatMode returns readOnly for PENDING order', () => {
  assert.equal(getPatientPharmacyChatMode({
    request: { requestType: 'CONSULTATION', chatRoomId: 'room-41' },
    order: { status: 'PENDING' },
  }), 'readOnly');
});

test('getPatientPharmacyChatMode returns hidden when chatRoomId is missing', () => {
  assert.equal(getPatientPharmacyChatMode({
    request: { requestType: 'CONSULTATION', status: 'IN_REVIEW', chatRoomId: null },
    order: null,
  }), 'hidden');
});

test('getPatientPharmacyChatMode returns hidden for PENDING request without order', () => {
  assert.equal(getPatientPharmacyChatMode({
    request: { requestType: 'CONSULTATION', status: 'PENDING', chatRoomId: 'room-41' },
    order: null,
  }), 'hidden');
});

test('getPatientPharmacyChatMode returns hidden for non-consultation request', () => {
  assert.equal(getPatientPharmacyChatMode({
    request: { requestType: 'ORDER_REQUEST', status: 'IN_REVIEW', chatRoomId: 'room-41' },
    order: null,
  }), 'hidden');
});

test('getPatientPharmacyChatMode returns hidden for falsy request', () => {
  assert.equal(getPatientPharmacyChatMode({ request: null, order: null }), 'hidden');
});

File: HealthLink_MB/test/pharmacy/chat/pharmacy_chat_policy_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:HealthLink/models/pharmacy/pharmacy_work_item.dart';
import 'package:HealthLink/utils/pharmacy/pharmacy_chat_policy.dart';

void main() {
  group('PharmacyChatPolicy.canEditWorkItem', () {
    PharmacyWorkItem _makeItem({
      String? requestType,
      String? chatRoomId,
      List<String> availableActions = const [],
    }) {
      return PharmacyWorkItem.fromJson({
        'workItemId': 'REQ-41',
        'sourceId': 41,
        'sourceType': 'REVISION',
        'workflowStage': 'REVISION_REQUESTED',
        if (requestType != null) 'requestType': requestType,
        if (chatRoomId != null) 'chatRoomId': chatRoomId,
        'availableActions': availableActions,
        'patientId': 'patient-1',
        'patientName': 'Patient',
        'createdAt': '2026-07-14T10:00:00',
      });
    }

    test('consultation with CHAT action and room is editable', () {
      final item = _makeItem(
        requestType: 'CONSULTATION',
        chatRoomId: 'room-41',
        availableActions: ['UPDATE_QUOTE', 'CHAT'],
      );
      expect(PharmacyChatPolicy.canEditWorkItem(item), isTrue);
    });

    test('missing chatRoomId is not editable', () {
      final item = _makeItem(
        requestType: 'CONSULTATION',
        availableActions: ['CHAT'],
      );
      expect(PharmacyChatPolicy.canEditWorkItem(item), isFalse);
    });

    test('non-consultation request type is not editable', () {
      final item = _makeItem(
        requestType: 'ORDER_REQUEST',
        chatRoomId: 'room-41',
        availableActions: ['CHAT'],
      );
      expect(PharmacyChatPolicy.canEditWorkItem(item), isFalse);
    });

    test('no CHAT action is not editable', () {
      final item = _makeItem(
        requestType: 'CONSULTATION',
        chatRoomId: 'room-41',
        availableActions: ['UPDATE_QUOTE'],
      );
      expect(PharmacyChatPolicy.canEditWorkItem(item), isFalse);
    });

    test('post-quote work item (no CHAT action) is not editable', () {
      final item = _makeItem(
        requestType: 'CONSULTATION',
        chatRoomId: 'room-41',
        availableActions: ['VIEW_ORDER'],
      );
      expect(PharmacyChatPolicy.canEditWorkItem(item), isFalse);
    });

    test('empty chatRoomId string is not editable', () {
      final item = _makeItem(
        requestType: 'CONSULTATION',
        chatRoomId: '',
        availableActions: ['CHAT'],
      );
      expect(PharmacyChatPolicy.canEditWorkItem(item), isFalse);
    });

    test('case insensitive requestType and CHAT action', () {
      final item = _makeItem(
        requestType: 'consultation',
        chatRoomId: 'room-41',
        availableActions: ['chat'],
      );
      expect(PharmacyChatPolicy.canEditWorkItem(item), isTrue);
    });
  });
}
